Make richer bootstrap data; use factories

weedySeaDragon opened the pull request:
@ERosendo  - Ready for an initial review

for #273 

Wanted to submit this as a draft before I go farther. (I still need to create Users that have auth to post.) (Excuse the 5,000 commits. I had to learn about mypy and its problems with Mocks.)

To keep responsibilities separate, I split the command from the class that makes the data:
  - `bootstrap-dev.py` has the command that calls the class to create the data
  - `MakeDevData` is the class that makes the data.  I named it this to parallel the name of the command used in CourtListener, but it might be better to name it "MakeBootstrapDevData" or something connected to the name of the command.

I created tests (`test_make_dev_data.py`) for MakeDevData both because I did some TDD and also to communicate what the class does (= a specification).

Here are some specifics:

### bootstrap-dev command
1. is now a subclass of VerboseCommand, which I copied from CourtListener
2. added these arguments (**_See the HELP screenshot below)_**
   1.  `--big-cases, -b` Number of subscriptions to make that are added to the big cases group. Default = **10**
   2. `--little-cases, -l` Number of subscriptions to make that are added to the big cases group. Default = **3**
   3. `--real-case, -r` A subscription created from a CourtListener docket; added to the big cases group.  Can specify multiple real cases.  Default = none
     _- I assumed it made sense to add it to the big cases group. Is this OK?_ 
3. Logs and prints (to stdout) the summary string that `MakeDevData().create()` returns after creating the data. (**_See the OUTPUT screenshot below_**)
5. added a method that takes a string and logs in (as `info`) and also prints it to stdout.
   - I've found this a really helpful utility and suggest that it go in the superclass (VerboseCommand). In the spirit of not doing things until they're needed, it also makes sense to leave it here and not refactor it into the superclass until there's another command that needs it

### MakeDevData
1. Always makes 1 admin user
1. Always makes the **big cases group** and 2 channels for it (Mastodon & Twitter)
1. Always makes the **little cases group** and 2 channels for it (Mastodon & Twitter)
2. Number of subscriptions made from random info = (number of big case subscriptions given + number of little case subscriptions given)
3. Number of subscriptions made from CL dockets = number of CL docket ids given


### UserFactory
- I added a ~~subclass to make an admin: AdminFactory~~ Trait to make an admin

## Here is the HELP (`--help`) for the bootstrap-dev specific arguments:


<img width="764" alt="bootstrap-dev HELP-2" src="https://github.com/freelawproject/bigcases2/assets/673794/28673363-269e-40b8-bf27-f6dd2d4c8fe7">

## Example OUTPUT (specified 2 big case subscriptions, 1 little case subscription, and 2 "real" cases from CL docket ids)

<img width="761" alt="bootstrap-dev EX-2" src="https://github.com/freelawproject/bigcases2/assets/673794/d70d1002-46e9-4c4c-bb54-2c17e2b90002">

mlissner commented:
Just checking in — is this one ready? It's still marked as draft?

weedySeaDragon commented:
I marked this as draft because I still need to add Users [per what I described in the comment of 273.](https://github.com/freelawproject/bigcases2/issues/273#issuecomment-1602726603)

I could do a separate PR for generating Users. Either way is fine with me.

And thanks for the reply re: reviewing PRs quickly.  I really appreciate that you like to keep the list of open PRs small. But I know that this piece is really low priority: it will be used rarely, and only be a few developers. :-)

ERosendo reviewed (changes requested):
This looks good. My initial suggestions are listed below:

weedySeaDragon commented on bc/users/tests/factories.py:
```diff
@@ -26,3 +26,9 @@ def channels(self, create, extracted, **kwargs):
         self.channels.add(*extracted)
+
+
+class AdminFactory(UserFactory):
```
I think just habit. (Other projects have more complex behavior in the Admin classes and so required more logic in the factory.)
Will definitely change.

weedySeaDragon commented on bc/core/management/commands/bootstrap-dev.py:
```diff
@@ -1,32 +1,80 @@
-                self.style.SUCCESS(f"Subscription: {subscription}")
-            )
+        real_cases = None
+        if options["real_cases"]:
```
Oops.  Nice catch.
(Thanks for helping me level up with python's ArguementParser :-) )

weedySeaDragon commented:
I made the requested changes
I also tried to improve the output: 
  - added a note to the command description that it ADDS data; it doesn't check for already existing data
  - improved the wording for the "Real subscriptions..." line
  - improved the wording for the "Subscriptions (random)..." line
  - "...subscribed to group..." is now an indented line; made a standard method for this. Created a mixin for the tests so it can be verified wherever it is used

I updated the `--help` and example output screenshots above.

weedySeaDragon commented:
I'll add the User data as a separate PR

ERosendo commented:
Thanks for this nice enhancement!.

mlissner commented:
Merged!
